Implemented the Node Alerter maintainer and metric.

mgimle opened the pull request:
This metric attempts to determine how many hosts it's possible to lose
before there's no place to fit its tenants, by finding a "shortest path to failure".
Since finding the actual path is np-hard, this maintainer constructs a
heuristic based on "repeated removals", and finds greedily finds a path
to failure with it.

The Node Alerter also exposes the "overcommittedNodes" metric, counting
how many hosts have children expecting more resources than it can provide.

Finally, this commit adds an obfuscated dump of data from zookeeper,
useful for running tests which require a node repository which reflects
reality.

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

bjorncs committed:
Simplify interface of vespa-curl-wrapper

bjorncs committed:
Use vespa-curl-wrapper in vespa-deploy

bjorncs committed:
Only generate Vespa TLS variables if client should use TLS

freva committed:
Move updateContainerNodeMetrics() to internal code

freva committed:
Non-functional cleanup

freva committed:
Hide current context from NodeAgent

bjorncs committed:
Use TlsContextBasedProvider in ConfigserverSslContextFactoryProvider

bjorncs committed:
Increase number of JRT transport threads

bjorncs committed:
Process all RPC requests in dedicated thread pool

- Rewrite unit tests to use a RPC client

bjorncs committed:
Create TestServer class that handle server side setup

Commit:
Nonfunctional changes only

Commit:
Nonfunctional changes only

Commit:
Nonfunctional changes only

Commit:
Block 1 ms, not 100 ms between cycles

freva committed:
Remove metrics check config writer code

arnej27959 committed:
rename dense::apply -> dense::generic_join

bjorncs committed:
Use current certificate manager to find certificate expiry

mgimle committed:
Implemented the Node Alerter maintainer and metric.
This metric attempts to determine how many hosts it's possible to lose
before there's no place to fit its tenants, by finding a "shortest path to failure".
Since finding the actual path is np-hard, this maintainer constructs a
heuristic based on "repeated removals", and finds greedily finds a path
to failure with it.

The Node Alerter also exposes the "overcommittedNodes" metric, counting
how many hosts have children expecting more resources than it can provide.

Finally, this commit adds an obfuscated dump of data from zookeeper,
useful for running tests which require a node repository which reflects
reality.

mgimle committed:
Implemented the Node Alerter maintainer and metric.
This metric attempts to determine how many hosts it's possible to lose
before there's no place to fit its tenants, by finding a "shortest path to failure".
Since finding the actual path is np-hard, this maintainer constructs a
heuristic based on "repeated removals", and finds greedily finds a path
to failure with it.

The Node Alerter also exposes the "overcommittedNodes" metric, counting
how many hosts have children expecting more resources than it can provide.

Finally, this commit adds an obfuscated dump of data from zookeeper,
useful for running tests which require a node repository which reflects
reality.

mgimle committed:
Merge branch 'master' into mgimle/docker-host-capacity-alert

bjorncs committed:
Revert "Use TlsContextBasedProvider in ConfigserverSslContextFactoryProvider"

This reverts commit e7cd527d1a83383188a65ed430f87b7fd9c85e60.

bjorncs committed:
Merge pull request #9966 from vespa-engine/bjorncs/revert

Revert "Use TlsContextBasedProvider in ConfigserverSslContextFactoryProvider"

bjorncs committed:
Use the jks keystore

bjorncs committed:
Merge pull request #9967 from vespa-engine/bjorncs/fix

Use the jks keystore

bjorncs committed:
Merge pull request #9952 from vespa-engine/bjorncs/config-proxy

Bjorncs/config proxy

Commit:
Add testrunner components only for application container clusters

bjorncs committed:
Remove unused parameter from constructor

bjorncs committed:
Specify correct file format for truststore

oyving commented on node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeAlerter.java:
```diff
@@ -0,0 +1,390 @@
+        HostRemovalFailure failureReason;
+    }
+
+    HostFailurePath worstCaseHostLossLeadingToFailure() {
```
Should return `Optional<HostFailurePath>` instead of null.

oyving commented on node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeAlerter.java:
```diff
@@ -0,0 +1,390 @@
+    HostFailurePath greedyHeuristicFindFailurePath(Map<Node, Integer> heuristic, List<Node> hosts,
+                                                   Map<Node, List<Node>> nodeChildren,
+                                                   Map<Node, AllocationResources> availableResources) {
+        if (hosts.size() == 0) return null;
```
See comment on `Optional<...>`.

oyving commented on node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/NodeAlerterTest.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    @Test
+    public void testWithGeneratedData() {
```
Split the different test cases you have in this method into multiple named tests

oyving reviewed (changes requested):
Not completely done with the review, but I think this is a good point to start fixing some minor things:
- The specific comments in the review.
- In general, there should be more JavaDoc minimally at the class level. But in the `NodeAlerter` class there is a lot of passing around of `Map<A, List<B>>` types that are difficult to follow. Some comments on the methods might help bring clarity here.
- Many of the methods can be marked as private, and quite a few of the inner classes can be marked as static. Inner classes should also have JavaDoc explaining what their intent is.

freva committed:
Merge pull request #9970 from vespa-engine/bjorncs/reapply-new-cfgserver-tls-setup

Bjorncs/reapply new cfgserver tls setup

havardpe committed:
Merge pull request #9963 from vespa-engine/arnej/rename-apply-to-join

rename dense::apply -> dense::generic_join

Commit:
Remove println

Commit:
Remove constraint that we can have only one main endpoint in EndpointList

Commit:
Merge pull request #9954 from vespa-engine/bratseth/dont-wait

Bratseth/dont wait

Commit:
Minor refactoring, move into separate package

freva committed:
Merge pull request #9959 from vespa-engine/freva/remove-metrics

Node-Admin: Remove node metrics related code

Commit:
Merge pull request #9971 from vespa-engine/ogronnesby/remove-endpoint-list-main-constraint

Remove constraint that we can have only one main endpoint in EndpointList

freva committed:
Merge pull request #9972 from vespa-engine/musum/minor-refactoring

Minor refactoring, move into separate package

Commit:
Merge pull request #9968 from vespa-engine/hmusum/add-testrunner-components-only-for-application-container-clusters

Add testrunner components only for application container clusters

mgimle committed:
Added more comments, made private members private,
made static inner class candidates static, moved filtering of uninteresting
hosts and tenants to the alerter from the test case generator, added
a test for overcommitted nodes, logically segmented the generated test cases,
removed explicit null checks in favour of Optionals, changed some inner
class initialization behaviour.

mgimle committed:
Added more comments, made private members private,
made static inner class candidates static, moved filtering of uninteresting
hosts and tenants to the alerter from the test case generator, added
a test for overcommitted nodes, logically segmented the generated test cases,
removed explicit null checks in favour of Optionals, changed some inner
class initialization behaviour.

bjorncs committed:
Expand VESPA_HOME variable

bjorncs committed:
Merge pull request #9973 from vespa-engine/bjorncs/tls-vespa-deploy

Bjorncs/tls vespa deploy

mgimle committed:
whitespace-caused merge conflict

mgimle commented:
Uh-oh, looks like my rebase caused some some weirdness. Trying again with another PR...
